// src/protos/env_manager.proto
syntax = "proto3";

package envmanager;

service EnvManagerService {
  rpc GetEnvStatus(GetEnvStatusRequest) returns (EnvStatusResponse);
  rpc CreateEnv(CreateEnvRequest) returns (CreateEnvResponse);
  rpc DeleteEnv(DeleteEnvRequest) returns (DeleteEnvResponse);

  rpc AddDeps(DepsRequest) returns (DepsOperationResponse);
  rpc UpdateDeps(DepsRequest) returns (DepsOperationResponse);
  rpc RemoveDeps(DepsRequest) returns (DepsOperationResponse);
  rpc ListDeps(ListDepsRequest) returns (ListDepsResponse);

  rpc SyncEnv(SyncEnvRequest) returns (SyncEnvResponse);
  rpc RunCode(RunCodeRequest) returns (RunCodeResponse);
  rpc ExportEnv(ExportEnvRequest) returns (ExportEnvResponse);
  rpc Cleanup(CleanupRequest) returns (CleanupResponse);

  rpc GetOperations(GetOperationsRequest) returns (OperationListResponse);
  rpc DeleteOperations(DeleteOperationsRequest) returns (DeleteOperationsResponse);
}

message GetEnvStatusRequest {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
}

message EnvStatusResponse {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string status = 4;
  string env_path = 5;
  bool has_pyproject = 6;
  bool has_uv_lock = 7;
  bool has_venv = 8;
  string metadata_json = 9;
}

message CreateEnvRequest {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string python_version = 4;
  repeated string packages = 5;
  bytes requirements_file_content = 6;
  string requirements_file_name = 7;
}

message CreateEnvResponse {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string env_path = 4;
  string python_version = 5;
  string status = 6;
  string pyproject_toml = 7;
}

message DeleteEnvRequest {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
}

message DeleteEnvResponse {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string status = 4;
}

message DepsRequest {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  repeated string packages = 4;
  bytes requirements_file_content = 5;
  string requirements_file_name = 6;
}

message DepsOperationResponse {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string status = 4;
  string stdout = 5;
  string stderr = 6;
  int32 exit_code = 7;
}

message ListDepsRequest {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
}

message ListDepsResponse {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  repeated string dependencies = 4;
  map<string, string> locked_versions = 5;
}

message SyncEnvRequest {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
}

message SyncEnvResponse {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string status = 4;
  string stdout = 5;
  string stderr = 6;
  int32 exit_code = 7;
}

message RunCodeRequest {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string code = 4;
  int32 timeout = 5;
}

message RunCodeResponse {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string stdout = 4;
  string stderr = 5;
  int32 exit_code = 6;
}

message ExportEnvRequest {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
}

message ExportEnvResponse {
  string workflow_id = 1;
  string node_id = 2;
  string version_id = 3;
  string pyproject_toml = 4;
  string uv_lock = 5;
}

message CleanupRequest {
  int32 idle_hours = 1;
}

message CleanupResponse {
  repeated string deleted = 1;
  int64 checked_at_unix = 2;
}

message GetOperationsRequest {
  string workflow_id = 1;
  string node_id = 2;
  string status = 3;
  int32 limit = 4;
  int32 offset = 5;
}

message OperationRecord {
  int64 id = 1;
  string workflow_id = 2;
  string node_id = 3;
  string version_id = 4;
  string operation = 5;
  string status = 6;
  int64 started_at_unix = 7;
  int64 finished_at_unix = 8;
  int64 duration_ms = 9;
  string stdout = 10;
  string stderr = 11;
  int32 exit_code = 12;
  string error = 13;
  string metadata_json = 14;
}

message OperationListResponse {
  int32 total = 1;
  int32 limit = 2;
  int32 offset = 3;
  repeated OperationRecord items = 4;
}

message DeleteOperationsRequest {
  string workflow_id = 1;
  string node_id = 2;
}

message DeleteOperationsResponse {
  string workflow_id = 1;
  int32 deleted_count = 2;
  string node_id = 3;
  repeated string deleted_node_ids = 4;
}